integrationtests/testground/main.go

```diff
@@ -0,0 +1,453 @@
+type SharedInfo struct {
+	Addr      string // multiaddr.Multiaddr
+	ThreadKey string // thread.Key
+}
+
+type threadWithKeys struct {
+	thread.Info
+	identity thread.Identity
+	logSk    crypto.PrivKey
+	logPk    crypto.PubKey
+	cli      *client.Client
+}
+
+func createThread(ctx context.Context, cli *client.Client) (thr *threadWithKeys, err error) {
+	// Create a thread, keeping read key and log private key on the client
+	id := thread.NewIDV1(thread.Raw, 32)
+	tk := thread.NewRandomKey()
+	logSk, logPk, err := crypto.GenerateEd25519Key(crand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	sk, _, err := crypto.GenerateEd25519Key(crand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	identity := thread.NewLibp2pIdentity(sk)
+	tok, err := cli.GetToken(ctx, identity)
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := cli.CreateThread(ctx, id,
+		corenet.WithThreadKey(tk),
+		corenet.WithLogKey(logPk),
+		corenet.WithNewThreadToken(tok))
+	if err != nil {
+		return nil, err
+	}
+	return &threadWithKeys{info, identity, logSk, logPk, cli}, nil
+}
+
+// joinThread joins to a thread created by another peer. It allows reading records created by the peer, also creates its own log to the thread.
+func joinThread(ctx context.Context, cli *client.Client, shared *SharedInfo) (*threadWithKeys, error) {
+	addr, err := multiaddr.NewMultiaddr(shared.Addr)
+	if err != nil {
+		return nil, err
+	}
+	key, err := thread.KeyFromString(shared.ThreadKey)
+	if err != nil {
+		return nil, err
+	}
+	logSk, logPk, err := crypto.GenerateEd25519Key(crand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	sk, _, err := crypto.GenerateEd25519Key(crand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	identity := thread.NewLibp2pIdentity(sk)
+	tok, err := cli.GetToken(ctx, identity)
+	if err != nil {
+		return nil, err
+	}
+	info, err := cli.AddThread(ctx, addr, corenet.WithThreadKey(key), corenet.WithLogKey(logSk), corenet.WithNewThreadToken(tok))
+	if err != nil {
+		return nil, err
+	}
+	return &threadWithKeys{info, identity, logSk, logPk, cli}, nil
+}
+
+func emptyThread(cli *client.Client) (thr *threadWithKeys) {
+	return &threadWithKeys{cli: cli}
+}
+
+func (t *threadWithKeys) Sharable() *SharedInfo {
+	return &SharedInfo{t.Addrs[0].String(), t.Key.String()}
+}
+
+// WaitForRecords blocks until it receives nRecords, then return them
+func (t *threadWithKeys) WaitForRecords(ctx context.Context, nRecords int) (records []corenet.Record) {
+	ch, err := t.cli.Subscribe(ctx, corenet.WithSubFilter(t.Info.ID))
+	if err != nil {
+		msg("Error subscribing thread: %v", err)
+		return
+	}
+	debug("Subscribed to thread")
+	total := 0
+	for record := range ch {
+		total++
+		records = append(records, record.Value())
+		if total == nRecords {
+			msg("Got all %d records", nRecords)
+			return
+		}
+	}
+	return
+}
+
+func (t *threadWithKeys) AddRecord(ctx context.Context, prev cid.Cid) (rec corenet.Record, err error) {
+	obj := make(map[string][]byte)
+	fuzzer.Fuzz(&obj)
```

Comment:
This is to create payload of different sizes / edge cases. Not meant to fuzz test cbor implementation. Just want to add some variance to the network traffic.